### ISSUE TITLE
On windows sending binary messages over data channel sometimes breaks

### ISSUE THREAD

**Issue by simonbengtsson** [flutter-webrtc/flutter-webrtc#908]
Sometimes when sending a binary message on windows it becomes corrupted. It was first noticed when sending an mp3 file which became corrupt when sending from windows but worked on all other platforms. After research it was found that sending bytes equal to zero doesn't work and causes the remaining part of the message to be corrupted.

**To Reproduce**
```dart
// Sending (from windows)
var bytes = Uint8List.fromList([5, 5, 0, 5, 5]);
var message = RTCDataChannelMessage.fromBinary(bytes);
dataChannel.send(message);

// Receiving (any platform including windows)
dataChannel.onMessage = (message) {
  print(message.binary); // Expected [5, 5, 0, 5, 5], actual [5, 5, 0, 200, 200] ie any bytes after the 0 turns into 200
};
```

**Workaround**
Send text message instead of binary. For example encoding bytes as base64.

```dart
import 'dart:convert';

// Sending
var bytes = Uint8List.fromList([5, 5, 0, 5, 5]);
var base64 = base64.encode(bytes);
var message = RTCDataChannelMessage(base64);
dataChannel.send(message);

// Receiving
dataChannel.onMessage = (message) {
  var bytes = base64.decode(message.text);
  print(bytes); // [5, 5, 0, 5, 5]
};
```

**Platform information**

* **Flutter version**: Flutter 2.12.0-4.1.pre • channel beta
* **Plugin version**: 0.8.5 (with patch for https://github.com/flutter-webrtc/flutter-webrtc/issues/907), not verified but likely also 0.8.3 
* **OS**: Windows 11

**Comment by pwinter1991** [comment]
I don't think that the sending of binary data is working on any platform. As the binary data is passed down to the other layers e.g. the webrtc wrapper as a string and the std::string constructor only reads till it finds a zero terminator. So everything after is not set correctly and contains garbage. Not even going to mention the obvious errors in the code like using const parameter that is copied by value.

```
void RTCDataChannelImpl::Send(const string data, bool binary /*= false*/) {
  if (binary) {
    rtc::CopyOnWriteBuffer binary(to_std_string(data));
    webrtc::DataBuffer buffer(binary, true);
    rtc_data_channel_->Send(buffer);
  } else {
    webrtc::DataBuffer buffer(to_std_string(data));
    rtc_data_channel_->Send(buffer);
  }
```

**Comment by waichungo** [comment]
Any progress on this issue?

**Comment by pwinter1991** [comment]
I still think the problem stems from passing the binary data through the layers encapsulated in a custom string class that gets converted to a std::string in the real send call on the data channel object.

**Comment by pwinter1991** [comment]
@cloudwebrtc 
Hey i opened 2 PRs that fix the data corruption issue with binary data. We can of course disuss how it really should be implemented but those 2 PRs are already manually tested to solve the problem. I'm not a big fan of the custom string class especially without specifying the namespace there is an automatically assumption that it would be std::string. It just doesn't make sense to put binary data in a class that handles concerns of strings where its pretty easy to lose the needed null bytes and instead getting memory garbage.

**Comment by pwinter1991** [comment]
Hey @simonbengtsson ,idk if you know but if the windows part currently blocks you could consume my repository through changing the dependency in the pubspec.yaml to, till the changes get merged:

```
flutter_webrtc:
    git:
      url: https://github.com/pwinter1991/flutter-webrtc
```

**Comment by simonbengtsson** [comment]
Fixed in 0.8.7
